### PR TITLE
Initial SQL 16 parser checkin

### PR DESCRIPTION
**Pull request by kaikalur**
This is the new javacc-based parser for parsing SQL 16 along with some of the Presto extensions.

### CODE REVIEW

File: .idea/.gitignore
Comment (rongrong) [inline .idea/.gitignore:2]: We should probably setup .gitignore properly at base dir.

File: parser/javacc-options-java.txt
Comment (rongrong) [inline parser/javacc-options-java.txt:6]: Should we just remove these?

File: parser/notes
Comment (rongrong) [inline parser/notes:9]: "pointer" like?

File: parser/notes
Comment (rongrong) [inline parser/notes:1]: Can you format this note to be more reader-friendly? I can hardly parse out how many different topics this is talking about...

File: parser/pom.xml
Comment (rongrong) [inline parser/pom.xml:43]: format

File: parser/pom.xml
Comment (rongrong) [inline parser/pom.xml:120]: format

File: parser/regexp.txt
Comment (rongrong) [inline parser/regexp.txt:2]: Is the regexp from spec? I think we allow different regexp libraries in presto?

**Review by rongrong — COMMENTED** [review]
Some general comments:
1) There are many commented out grammars, why are the commented out? Should we either provide a reason or remove them?
2) There are a lot of TODO(sreeni) maybe you want to change to TODO(githubId) instead?
3) might be nicer to separate the grammar, Java and C++ into different commits
4) We should at least at some tests to this before merging.
5) Should update the .gitignore to ignore .idea
6) is javacc.jar supposed to be merged here?

File: .idea/.gitignore
Comment (kaikalur) [reply .idea/.gitignore:2]: Done. Copied from Presto

File: parser/javacc-options-java.txt
Comment (kaikalur) [reply parser/javacc-options-java.txt:6]: Done

File: parser/notes
Comment (kaikalur) [reply parser/notes:9]: Done

File: parser/notes
Comment (kaikalur) [reply parser/notes:1]: Done

File: parser/pom.xml
Comment (kaikalur) [reply parser/pom.xml:43]: Done

File: parser/pom.xml
Comment (kaikalur) [reply parser/pom.xml:120]: Done

File: parser/regexp.txt
Comment (kaikalur) [reply parser/regexp.txt:2]: We simply use a string literal for regex. But the spec actually has a full regex spec in the grammar! We have that fragment of the grammar in the repo as well if we want to eventually support that.

**Comment by kaikalur** [comment]
> Some general comments:
> 
> 1. There are many commented out grammars, why are the commented out? Should we either provide a reason or remove them?

So these are for things that the spec actually has subgrammar for (like date literals and regex). So I just left them in there commented out so it's readable.

> 2. There are a lot of TODO(sreeni) maybe you want to change to TODO(githubId) instead?

Done.

> 3. might be nicer to separate the grammar, Java and C++ into different commits

Will try 👍 

> 4. We should at least at some tests to this before merging.
> 5. Should update the .gitignore to ignore .idea

Done

> 6. is javacc.jar supposed to be merged here?

Yes - we don't have a good maven plugin. I will research more later and fix it.

**Comment by kaikalur** [comment]
Addressed all the first of comments - 3 different commits, a simple smoke test and other cleanup.

File: parser/cpp/tokenize.cc
Comment (rongrong) [inline parser/cpp/tokenize.cc:50]: Why are these commented out?

File: parser/cpp/tokenize.cc
Comment (rongrong) [inline parser/cpp/tokenize.cc:56]: It's been too long since I last saw this. Why is `#if 0` useful?

File: parser/kw.txt
Comment (rongrong) [inline parser/kw.txt:125]: Why is this commented out?

File: parser/kw.txt
Comment (rongrong) [inline parser/kw.txt:377]: I think the ones commented are made non-reserved against the spec. Since these are already mentioned above, I think you can just remove them here. Or if you prefer to keep the original reserved list, maybe add a comment to explain this.

File: parser/notes
Comment (rongrong) [reply parser/notes:1]: Still didn't quite get what's the purpose of the note. Could they go into the code? What do you want to achieve but putting them in notes here?

File: parser/regexp.txt
Comment (rongrong) [reply parser/regexp.txt:2]: Hmm what does "unsupported" here mean? CoreSQL not supporting or Presto? There are probably more stuff that Presto doesn't support?

File: parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java
Comment (rongrong) [inline parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java:25]: Can we change the test to read from an input file or something so we can add more test cases as text?

File: src/modernizer/violations.xml
Comment (rongrong) [inline src/modernizer/violations.xml:1]: Do we need this?

File: src/spotbugs/spotbugs-exclude.xml
Comment (rongrong) [inline src/spotbugs/spotbugs-exclude.xml:1]: Do we need this?

**Review by rongrong — APPROVED** [review]
The parser/cpp/token_manager.h seems to be empty. Please remove. I'd also prefer to not include the javacc.jar in the source code if possible.

File: parser/notes
Comment (kaikalur) [reply parser/notes:1]: > Still didn't quite get what's the purpose of the note. Could they go into the code? What do you want to achieve but putting them in notes here?

These are explaining some things that are specific to Javacc and why we did it that way. Part of it is for giving pointers to future maintainers on how to do things.

File: parser/regexp.txt
Comment (kaikalur) [reply parser/regexp.txt:2]: > Hmm what does "unsupported" here mean? CoreSQL not supporting or Presto? There are probably more stuff that Presto doesn't support?

I talk about it in the notes now. Basically, any syntax that we don't support right now turns into an 'Unsupported' AST node so that we can give good error messages in the right places.

File: parser/kw.txt
Comment (kaikalur) [reply parser/kw.txt:125]: Done

File: parser/kw.txt
Comment (kaikalur) [reply parser/kw.txt:377]: Done

File: parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java
Comment (kaikalur) [reply parser/src/test/java/com/facebook/coresql/parser/TestSqlParser.java:25]: Yeah - maybe next round.

File: src/modernizer/violations.xml
Comment (kaikalur) [reply src/modernizer/violations.xml:1]: Removed

File: src/spotbugs/spotbugs-exclude.xml
Comment (kaikalur) [reply src/spotbugs/spotbugs-exclude.xml:1]: Removed

**Comment by kaikalur** [comment]
Cleaned up the cpp version as well. Just do ./build.sh to build the cpp version now which uses maven plugin to generate the C++ parser as well.

File: parser/notes
Comment (rongrong) [reply parser/notes]: So for people who understands javacc they will know what you mean?

File: parser/grammar/presto-extensions.txt
Comment (rongrong) [inline parser/grammar/presto-extensions.txt:109]: Wait does this mean all aggregation functions have to be defined in grammar?

File: parser/pom.xml
Comment (rongrong) [inline parser/pom.xml:21]: nits: remove empty line

File: parser/pom.xml
Comment (rongrong) [inline parser/pom.xml:37]: nits: remove empty line.

File: parser/pom.xml
Comment (rongrong) [inline parser/pom.xml:103]: nits: Can we be consistent with empty lines? Either have them or not between items :P

File: pom.xml
Comment (rongrong) [inline pom.xml:33]: empty line in between?

File: parser/cpp/javacc-options.txt
Comment (rongrong) [reply parser/cpp/javacc-options.txt:9]: These are still here :P

File: pom.xml
Comment (rongrong) [inline pom.xml:22]: Can we remove the `<dependencies>` from previous commit?

File: parser/cpp/javacc-options.txt
Comment (kaikalur) [reply parser/cpp/javacc-options.txt]: Done

File: pom.xml
Comment (kaikalur) [reply pom.xml:28]: Removed all these

File: pom.xml
Comment (rongrong) [inline pom.xml]: Remove this in the second commit.

File: parser/grammar/nonreservedwords.txt
Comment (rongrong) [inline parser/grammar/nonreservedwords.txt:5]: Does the ones that are reserved words in spec but we changed to non-reserved belong here too?

**Review by rongrong — APPROVED** [review]

File: parser/grammar/kw.txt
Comment (kaikalur) [reply parser/grammar/kw.txt]: Done for real!

**Pull request merged**
